portal/server/controllers/utilisation-report-service/utilisation-report-upload/utilisation-report-validator.js

```diff
@@ -82,8 +82,60 @@ const validateCsvData = (csvData) => {
   return validationErrors;
 };
 
+/**
+ * Checks the filename parameter to see if it contains a month anywhere
+ * in the filename and, if it does, returns an object containing both
+ * variations of the month name. If it doesn't find any months, it will
+ * return undefined
+ * @param {string} filename - The filename
+ * @returns {{ longName: string, shortName: string } | undefined}
+ */
+const getMonthInFilename = (filename) => {
+  const lowerCaseFilename = filename.toLowerCase();
+  const monthNames = Object.values(LOWER_CASE_MONTH_NAMES).find(
+    (month) => lowerCaseFilename.includes(month.longName) || lowerCaseFilename.includes(month.shortName),
+  );
+  return monthNames;
+};
+
+/**
+ * Given a filename and a report period, this function checks whether
+ * or not the filename contains a month and, if it does, whether or
+ * not the month matches the report period passed in. If it does match,
+ * it returns an empty object. Otherwise, it returns an object containing
+ * a filename error message related to the specific case reached
+ * @param {string} filename - The filename, using either '_' or '-' as a separator
+ * @param {string} dueReportPeriod - The current due report period with format 'MMMM yyyy'
+ * @returns {{ filenameError: string | undefined }}
+ */
+const validateFilenameContainsReportPeriod = (filename, dueReportPeriod) => {
+  const exampleFilenameReportPeriod = dueReportPeriod.replace(' ', '_');
+
+  const monthInFilename = getMonthInFilename(filename);
+  if (!monthInFilename) {
+    const filenameError = `The selected file must contain the reporting period as part of its name, for example '${exampleFilenameReportPeriod}'`;
+    return { filenameError };
+  }
+
+  const currentReportPeriodYear = dueReportPeriod.split(' ').at(-1);
+  const longFilenameReportPeriod = `${monthInFilename.longName}_${currentReportPeriodYear}`;
+  const shortFilenameReportPeriod = `${monthInFilename.shortName}_${currentReportPeriodYear}`;
+  const lowerCaseFilenameWithUnderscores = filename.toLowerCase().replaceAll('-', '_');
+  if (lowerCaseFilenameWithUnderscores.includes(longFilenameReportPeriod)) {
+    return {};
+  }
+  if (lowerCaseFilenameWithUnderscores.includes(shortFilenameReportPeriod)) {
+    return {};
+  }
```

Comment:
As discussed, I think we could do this with a regex.

Reply:
Done